### DISCUSSION TITLE
Apparent limitation of using quickUMLS with Spacy ner disabled

### DISCUSSION THREAD

**Discussion by RonWilkinson** [medspacy/medspacy#67] [Q&A] [answered]
In order to get quickUMLS to successfully annotate CUI concepts, I had to disable Spacy's default ner named entity recognizer.  However, ner seems to be good at finding quantities such as "15 kg", percentages like "98%", and cardinal numbers as in "7 treatments", which are not detected when ner is omitted from the pipeline.   Is there an easy alternate way to add these numeric detections back in without inhibiting quickUMLS?

**Comment by turbosheep** [comment, marked as answer]
Do you have a small example you could provide that shows the output you receive versus the output you would expect? This sounds like it might be unintended behavior on the part of QuickUMLS if you get NO output from quickUMLS when you have previous entity spans in the text. 

While spaCy does not allow spans in `doc.ents` to overlap, my experience is that the default NER components generally have little overlap with other components of the pipeline, so you should be getting at least some results from both components.

In the meantime, a few possible solutions you could try:

1. Adding QuickUMLS before the spacy NER component in the pipeline. This way the CUIs produced by QuickUMLS will take precedence while afterwards the NER can add in additional spans. This might also avoid the possible unintentional behavior where QuickUMLS does not add spans if a prior NER component is present. Something like this would work:
```python
nlp = spacy.load("en_core_web_sm")
_, ner = nlp.remove_pipe("ner")

quickumls = QuickUMLS(...)

nlp.add_pipe(quickumls)
nlp.add_pipe(ner)

print(nlp.pipe_names)
>>> ['tagger', 'parser', 'quickumls', 'ner']
```

2. If all you are interested from the NER output is measurements, quantities, counts, etc., a simple [EntityRuler](https://v2.spacy.io/api/entityruler) from default spacy or the TargetMatcher in medspacy could help. A spacy pattern something like `[{"LIKE_NUM": True}, {"LOWER": {"IN": ["mg", "kg", ...]}}]` could capture units of measure for example for any number followed by a unit of measure you specify. A few rules could cover a large number of possible outputs of that format with any label you find appropriate, whether that is matching the label from spacy's NER or whether it is something you specify.

**Reply by RonWilkinson** [reply]
I tested suggestion 1. 

1. I did a fresh boot of the jupyter notebook server.
2.  I created a new notebook.
3. I created a spacy pipeline using the steps in the suggestion.  (The only tweak needed was to use the syntax: nlp.add_pipe(ner[1]), because ner = nlp.remove_pipe("ner") returns a tuple ("ner", <ner component>) and add_pipe just wants the component.)
4. Both quickumls and ner returned named entities as desired. Great!

However, I found that if ner appears even once  in a pipeline before quickumls, quickumls will not work for any new pipelines with or without ner or in any new notebooks at all until the jupyter notebook server is shut down and restarted from scratch.

This is all on the Windows platform.

**Reply by RonWilkinson** [reply]
In line three, ....returns a tuple("ner", \<ner component\>, and the add_pipe just wants the component.
